### DISCUSSION TITLE
Can we use swift-url-routing with WebSocket?

### DISCUSSION THREAD

**Discussion by mackoj** [pointfreeco/swift-url-routing#16] [Q&A] [answered]
Hi,

Can we use swift-url-routing with WebSocket? If not do you plan to add support for it ?

Thanks,

**Comment by stephencelis** [comment, marked as answer]
We haven't thought through the details here, but could be interesting! Wanna sketch out what you're thinking?

**Reply by mackoj** [reply]
> Wanna sketch out what you're thinking?

I don't have a full plan to make it work yet but I will try 👍

Thanks,

**Reply by mackoj** [reply]
The basic idea for the client part is to:
- make a wrapper for `URLSessionWebSocketTask` that conform to `AsyncStream` let's call it **WebSocketTask**
- then we need a **WebSocketTask** consumer that is a `ParserPrinter` let's call it **WebsocketRoute**
- **WebsocketRoute** will listen to the `Path` if it match then for each message it receive it call a `ParserPrinter` called **Messages** that will try to consume Websocket.Message that it received
- **Messages** parsers will just transform `Data` -> T 

```swift
public let mainRouter = OneOf {
   WebsocketRoute {
       Path { "sse" }
       Message {
           Body(.json(Message1.self))
           Body(.json(Message2.self))
           Body(.json(Message3.self))
           Body(.json(Message4.self))
           Body(.json(Message5.self))
       }
   }
}
```

I think most of the pieces are here for the client part. What do you think ?

Cheers,

**Reply by mackoj** [reply]
With a few modifications this could work for Bonjour(Zeroconf) too 🤔

**Reply by alexataman** [reply]
Hi @mackoj, did you manage to do that? I'm working on a similar task right now and trying to find the solution

**Reply by mackoj** [reply]
Hi @alexataman I did take another route to support WebSocket in my app sorry.

**Reply by alexataman** [reply]
Gotcha, thanks for replying
